src/tests/Microsoft.PowerFx.Json.Tests.Shared/AsTypeIsTypeParseJSONTests.cs

```diff
@@ -0,0 +1,200 @@
+        }
+
+        [Fact]
+        public void TablesTest()
+        {
+            var engine = SetupEngine();
+
+            engine.AddUserDefinitions("T = Type([{a: Number}]);");
+
+            var t1 = new object[] { 5D };
+            var t2 = new object[] { 1m, 2m, 3m, 4m };
+            var t3a = new object[] { true, true, false, true };
+            var t3 = new object[] { t3a };
+
+            CheckIsTypeAsTypeParseJSON(engine, "\"[{\"\"a\"\": 5}]\"", "T", t1);
+            CheckIsTypeAsTypeParseJSON(engine, "\"[{\"\"a\"\": 5}]\"", "Type([{a: Number}])", t1, options: ParseType);
+            CheckIsTypeAsTypeParseJSON(engine, "\"[{\"\"a\"\": [true, true, false, true]}]\"", "Type([{a: [Boolean]}])", t3, options: ParseType);
+            CheckIsTypeAsTypeParseJSON(engine, "\"[1, 2, 3, 4]\"", "Type([Decimal])", t2, options: ParseType);
+
+            // Negative tests
+            CheckIsTypeAsTypeParseJSON(engine, "\"[{\"\"a\"\": 5, \"\"b\"\": 6}]\"", "Type([{a: Number}])", t1, false, options: ParseType);
+            CheckIsTypeAsTypeParseJSON(engine, "\"[1, 2, 3, 4]\"", "Type([Text])", t2, false, options: ParseType);
+            CheckIsTypeAsTypeParseJSONCompileErrors(engine, "\"[\"\"foo/bar/uri\"\"]\"", "Type([Color])", TexlStrings.ErrUnsupportedTypeInTypeArgument.Key, options: ParseType);
+        }
+
+        [Theory]
+        [InlineData("\"42\"", "SomeType", true, "ErrInvalidName")]
+        [InlineData("\"42\"", "Type(5)", true, "ErrTypeLiteral_InvalidTypeDefinition")]
+        [InlineData("\"42\"", "Text(42)", true, "ErrInvalidArgumentExpectedType")]
+        [InlineData("\"\"\"Hello\"\"\"", "\"Hello\"", true, "ErrInvalidArgumentExpectedType")]
+        [InlineData("\"{}\"", "Type([{a: 42}])", true, "ErrTypeLiteral_InvalidTypeDefinition")]
+        [InlineData("AsType(ParseJSON(\"42\"))", "", false, "ErrBadArity")]
+        [InlineData("IsType(ParseJSON(\"42\"))", "", false, "ErrBadArity")]
+        [InlineData("AsType(ParseJSON(\"42\"), Number , Text(5))", "", false, "ErrBadArity")]
+        [InlineData("IsType(ParseJSON(\"42\"), Number, 5)", "", false, "ErrBadArity")]
+        [InlineData("AsType(ParseJSON(\"123\"), 1)", "", false, "ErrInvalidArgumentExpectedType")]
+        public void TestCompileErrors(string expression, string type, bool testAllFunctions, string expectedError)
+        {
+            var engine = SetupEngine();
+
+            if (testAllFunctions)
+            {
+                CheckIsTypeAsTypeParseJSONCompileErrors(engine, expression, type, expectedError, ParseType);
+            }
+            else
+            {
+                var result = engine.Check(expression, options: ParseType);
+                Assert.False(result.IsSuccess);
+                Assert.Contains(result.Errors, e => e.MessageKey == expectedError);
+            }
+        }
+
+        [Fact]
+        public void TestFunctionsWithTypeArgs()
```

Comment:
>TestFunctionsWithTypeArgs [](http://example.com/codeflow?start=20&length=25)

It would be also good to assert that for functions that do **not** have type args, then none of ArgIsType returns true. Something like

    var functionsWithoutTypeArgs = TestOnly_AllBuiltinFunctions.Where(f => !f.HasTypeArgs);
    Assert.All(functionsWithoutTypeArgs, f => !Enumerable.Range(0, Math.Min(f.MaxArity, 5)).Any(index => f.ArgIsType(index)))